src/registrar/models/domain.py

```diff
@@ -1648,6 +1667,105 @@ def _get_dnssec_data(self, response_extensions):
+        old_cache_hosts = self._cache.get("hosts")
+        old_cache_contacts = self._cache.get("contacts")
+
+        if fetch_contacts:
+            cleaned["contacts"] = self._get_contacts(cleaned.get("_contacts", []))
+            if old_cache_hosts is not None:
+                logger.debug("resetting cleaned['hosts'] to old_cache_hosts")
+                cleaned["hosts"] = old_cache_hosts
+
+        if fetch_hosts:
+            cleaned["hosts"] = self._get_hosts(cleaned.get("_hosts", []))
+            if old_cache_contacts is not None:
+                cleaned["contacts"] = old_cache_contacts
+
+    def _update_hosts_and_ips_in_db(self, cleaned, fetch_hosts):
+        """Update hosts and host_ips in database if retrieved from registry.
+
+        Parameters:
+            self: the domain to be updated with hosts and ips from cleaned
+            cleaned: dict containing hosts.  Hosts are provided as a list of dicts, e.g.
+                [{"name": "ns1.example.com",}, {"name": "ns1.example.gov"}, "addrs": ["0.0.0.0"])]
+            fetch_hosts: boolean indicating whether or not fetch_hosts was called
+        """
+        if fetch_hosts:
+            cleaned_hosts = cleaned["hosts"]
+            # Get all existing hosts from the database for this domain
+            existing_hosts_in_db = Host.objects.filter(domain=self)
+            # Identify hosts to delete
+            cleaned_host_names = set(cleaned_host["name"] for cleaned_host in cleaned_hosts)
+            hosts_to_delete_from_db = [
+                existing_host for existing_host in existing_hosts_in_db if existing_host.name not in cleaned_host_names
+            ]
+            # Delete hosts and their associated HostIP instances
+            for host_to_delete in hosts_to_delete_from_db:
+                # Delete associated HostIP instances
+                HostIP.objects.filter(host=host_to_delete).delete()
+                # Delete the host itself
+                host_to_delete.delete()
+            # Update or create Hosts and HostIPs
+            for cleaned_host in cleaned_hosts:
+                # Check if the cleaned_host already exists
+                host_in_db, host_created = Host.objects.get_or_create(domain=self, name=cleaned_host["name"])
+                # Get cleaned list of ips for update
+                cleaned_ips = cleaned_host["addrs"]
+                if not host_created:
+                    # Get all existing ips from the database for this host
+                    existing_ips_in_db = HostIP.objects.filter(host=host_in_db)
+                    # Identify IPs to delete
+                    ips_to_delete_from_db = [
+                        existing_ip for existing_ip in existing_ips_in_db if existing_ip.address not in cleaned_ips
+                    ]
+                    # Delete IPs
+                    for ip_to_delete in ips_to_delete_from_db:
+                        # Delete the ip
+                        ip_to_delete.delete()
```

Comment:
Something similar to above, but you can definitely get a list of existing_ip.id here (existing_ip.id for existing_ip, etc etc). By using `HostIP.objects.filter(id__in=ip_ids_to_delete_from_db).delete()`, you can get rid of the second for loop. That should be more computationally efficient

Reply:
this is on a different PR already merged, resolving